### ISSUE TITLE
Remove NLTK dependency

### ISSUE THREAD

**Issue by feanil** [openedx/openedx-chem#85]
An LMS web worker process uses over 300 MB of RAM once initialized. Over 10% of that is from loading [nltk 1](https://www.nltk.org/), which we only use in one place–to parse chemical equation inputs in ProblemBlocks (see [chemcalc.py 1](https://github.com/openedx/openedx-chem/blob/25ea1357a5bd1a58a205ceb3ea636a54449ee6ea/chem/chemcalc.py) in the openedx-chem repo).

It's not clear why the grammar is specified using NLTK instead of pyparsing. It could have been to get around some limitation that pyparsing had twelve years ago, or it could have just been that the author was just more familiar with NLTK and could hack the code out faster that way.

Task: Remove our dependency on NLTK by changing the parser implementation in this repo? It would likely involve some digging, and exact backwards compatibility would be extremely important.

More details about the process of getting this info: https://discuss.openedx.org/t/reducing-memory-usage-nltk/13406

**Comment by OmarIthawi** [comment]
@feanil @ormsbee The stop-gap solution for this fix is to lazyload nltk. Additionally, we could explore turning off this feature by default.

I don't think this is a very popular feature outside edX.org and MITx (maybe).

**Comment by feanil** [comment]
@OmarIthawi in this case it's being used to parse the content of the Chem Equation.  Can you provide more context on your approach?  I think you're saying, only import nltk and create a parser when we need to parse the content instead of globally when the package is loaded?

**Comment by OmarIthawi** [comment]
Yes @feanil, that's the easiest way to fix the RAM usage for the 99% of requests which doesn't use nltk.

It will also avoid the slow path of verification and backward compatibility, which can be very painful in a huge system like the edx-platform.

However, to me, I think we should move this feature into a more plugin instead of standard dependency so it's only loaded when needed. This way, it's no longer a "core" concern but an optional feature that comes with its own baggage. Which is totally fine imho.

We've done a good job by extracting it into its own module, another step is also useful.

btw, The Python PEPers have an interesting proposal: https://peps.python.org/pep-0810/

> lazy from json import dumps


What happens when the module is loaded in a `wsgi` process is what I'm unsure of. Would it put the same overhead, or the process is terminated periodically. I don't have the required understanding of how threads / processes / module loading works in Python at the moment.

**Comment by feanil** [comment]
Yea, moving it into its own xblock makes a lot of sense. Lazy importing would also solve this but given that they're targetting 3.15 for that, it would be nice to get something in sooner.

**Comment by OmarIthawi** [comment]
@feanil Lazy loading as a native python feature needs 3.15. But we can do it earlier by mimicking the lazy loading in the form of:

```
def nltk_module():
    import nltk
    return nltk
```

Then:

```diff
-import nltk
+from .utils import nltk_module

 def some_function_()
-    nltk.do_something()
+    nltk_module().do_something()
```

This is effectivly lazy loading that's very similar in benefits to the 3.15 proposal. Except in minor ways such as triggering a warning in pylint and so.

What do you think?

**Comment by feanil** [comment]
That seems fine given the benefit and we can just put the reasoning into a comment in the little functions.
